Improve custom intallation docs and add a new example (#38)
rancher/elemental-docs: docs/customizing.md

## Customize installation ISO and installation process

In order to adapt the installation ISO a simple approach is to append extra configuration
files into the ISO root in an analog way the registration yaml configuration file
is added.

### Common customization pattern

Elemental Teal installation can be customized in three different non-exclusive ways. First, including
some custom Elemental client configuration file, second, by including additional cloud-init files to execute at
boot time, and finally, by including installation hooks.

A common pattern is a combination of all three, this way custom steps can be added during the installation
and additional cloud-init files can be added to the installed system so they are evaluated at boot time.

> For that to happen the ISO needs to include:

Author: Thanks! :+1:

1. A [configuration file](https://rancher.github.io/elemental-toolkit/docs/customizing/general_configuration/)
   for the elemental client, at least to point installation hooks location. Commonly this is done by adding
   a `/elemental/config.yaml` file into the ISO root.

2. The additional cloud-init files to be included into the installed system. They
   allow to perform custom operations at boot time.

3. The installation hooks evalutated at install time. They allow to perform custom operations
   during the installation process (include extra software, set additional disks...).

#### Custom Elemental client configuration file

[Elemental client](https://github.com/rancher/elemental-cli) `install`, `upgrade` and `reset` commands can be configured with a
custom [configuration file](https://rancher.github.io/elemental-toolkit/docs/customizing/general_configuration/).

In order to set a custom configuration file the installation ISO must include it.
By default, the `elemental-register` command will attempt to load the Elemental client

> configuration within the `elemental` folder in side the ISO root. That is ISOs including

Reviewer: typo:
```suggestion
configuration within the `elemental` folder inside the ISO root. That is ISOs including
```

> `/elemental/config.yaml` file and/or multiple yaml files inside the `/elemental.conf.d` folder.

Author: Manually applied, thanks :+1:

Note the installation hooks only cover installation procedures, for upgrades equivalent
`before-upgrade` and/or `after-upgrade-chroot` should be defined.

### Adding extra LVM volume group disks during the installation

This example is covering the case the host has multiple disks, some of them used
as part of an LVM setup.

> In this particular case we will assume a host includes three disks (`/dev/sda`, `/dev/sdb`

Reviewer: We use vdX in the examples (see below comment)
```suggestion
In this particular case we will assume a host includes three disks (`/dev/vda`, `/dev/vdb`
```

Author: Already changed to `sd*` instead of `vd*`. I prefer using device names not suggesting libvirt use.

> and `/dev/sdc`), where the first one is used for a regular Elemental Teal installation

Reviewer: ```suggestion
and `/dev/vdc`), where the first one is used for a regular Elemental Teal installation
```

are created, formatted and mounted at boot time via an extended fstab file.

For that use case the following files are required:

* additional clout-init files to include within the installed system
* additional installation hooks to prepare the LVM volumes at install time
* additional Elemental client configuration file to point hooks file location

Lets create an `overlay` directory to create the directory tree that needs to be
added into the ISO root. In that case the `overlay` directory could contain:

```yaml showLineNumbers
overlay/
  oem/
    lvm_volumes_in_fstab.yaml
  hooks/
    lvm_volumes_hook.yaml
  elemental/
    config.yaml
```

The Elemental client config file in `overlay/elemental` could be as:

```yaml showLineNumbers
cloud-init-paths:
  - "/run/initramfs/live/hooks"
```

This is just to let Elemental client know where to find installation hooks.

The installation hook `overlay/hooks/lvm_volumes_hook.yaml` could be as:

```yaml showLineNumbers
name: "Create LVM logic volumes over some physical disks"
stages:
  post-install:
    - name: "Create physical volume, volume group and logical volumes"

>       if: '[ -e "/dev/vdb" ] && [ -e "/dev/vdc" ]'

Reviewer: above we assume /dev/sda, /dev/sdb and /dev/sdc . Here we are using /dev/vdb and /dev/vdc

Author: oh you are right 😓 Yes I did the tests under KVM and used virtio devices.

      commands:
      - | 
        # Create the physical volume, volume group and logical volumes
        pvcreate /dev/vdb /dev/vdc
        vgcreate elementalLVM /dev/vdb /dev/vdc
        lvcreate -L 8G -n elementalVol1 elementalLVM
        lvcreate -l 100%FREE -n elementalVol2 elementalLVM

        # Trigger udev detection
        if [ ! -e "/dev/elementalLVM/elementalVol1" ] || [ ! -e "/dev/elementalLVM/elementalVol2" ]; then
          sleep 10
          udeadm settle
        fi

        # Ensure devices are already available
        [ -e "/dev/elementalLVM/elementalVol1" ] || exit 1
        [ -e "/dev/elementalLVM/elementalVol2" ] || exit 1

        # Format logical volumes with a known label for later use in fstab
        mkfs.xfs -L eVol1 /dev/elementalLVM/elementalVol1
        mkfs.xfs -L eVol2 /dev/elementalLVM/elementalVol2
```

> It essentially creates and formats the LVM devices as desired. This is a good
> example of an installation hook, as this setup is only needed once at installation
> time. Doing such a thing on first boot is also a possibility, but it would definitely
> require a more sophisticated logic to ensure its only applied once at first boot.

Reviewer 2: ```suggestion
The LVM devices are created and formatted as desired. This is a good
example of an installation hook, as this setup is only needed once, at installation
time. As an alternative, the same action could be done on first boot, however it would
require a more sophisticated logic to ensure it's only applied once at first boot.
```

> Finally, the boot time cloud-init files have only to take care of setting the mount
> points and actually mounting at boot. In Elemental OS fstab is ephemeral, so there is
> no fstab set during the installation, it is dynamically created at boot time. So that
> it can't be set within an installation hook and a boot time cloud-init file is used.

Reviewer 2: ```suggestion
Finally, the boot time `cloud-init` files contain the mount points settings and trigger the
action of mounting those mountpoints. The Elemental OS `fstab` file is ephemeral and it's 
dynamically created at boot time. That's why it doesn't exist during the installation and
can't be used in an installation hook.
```

> The `overlay/oem/lvm_volumes_hook.yaml` could something as simple as:

Reviewer 2: ```suggestion
Here's an example of `overlay/oem/lvm_volumes_hook.yaml`:
```

```yaml showLineNumbers
name: "Mount LVM volumes"
stages:
  initramfs:
    - name: "Extend fstab to mount LVM logical volumes at boot"
      commands:
      - |
        echo "LABEL=eVol1 /usr/local/eVol1  xfs defaults  0 0" >> /etc/fstab
        echo "LABEL=eVol2 /usr/local/eVol2  xfs defaults  0 0" >> /etc/fstab
```

> Note the `initramfs` stage is the last stage inside the initfamfs right before
> switching root to the actual root tree, in runs chrooted to the final root. At this
> stage `/etc/fstab` already exists and it is a good place to adapt it, as after
> switching root systemd will handle the rest of the initalizaton and process apply it.

Reviewer 2: ```suggestion
:::note
The `initramfs` stage is the last stage before switching to the actual root tree.
At this stage, the `/etc/fstab` file already exists and can be adapted before
switching root. Once running in the final root tree, SystemD will handle the rest of the initialization and apply it.
:::
```

> This cloud-init file should be included into the `/oem` folder on the installed
> system. `/oem` is mount point for the OEM partition. In order include extra files
> there they should be listed as `config-urls` within the Registration CRD at the
> management cluster.

Reviewer 2: ```suggestion
This cloud-init file should be included into the `/oem` directory on the installed
system. `/oem` is a mount point for the OEM partition. In order to include extra files,
they should be listed as `config-urls` within the Registration CRD at the
management cluster.
```

### Repacking the ISO image with extra files

> ¨Assuming an `overlay` folder was created in the current directory containing all

Reviewer: guess a typo?

```suggestion
Assuming an `overlay` folder was created in the current directory containing all
```

> ¨Assuming an `overlay` folder was created in the current directory containing all

Author: yes it was, already fixed, thanks
